service/debugger/locations.go

```diff
@@ -0,0 +1,291 @@
+
+	case '/':
+		rx, rest := readRegex(rest[1:])
+		if len(rest) < 0 {
+			return nil, malformed("non-terminated regular expression")
+		}
+		if len(rest) > 1 {
+			return nil, malformed("no line offset can be specified for regular expression locations")
+		}
+		return &RegexLocationSpec{rx}, nil
+
+	case '*':
+		rest = rest[1:]
+		addr, err := strconv.ParseInt(rest, 0, 64)
+		if err != nil {
+			return nil, malformed(err.Error())
+		}
+		if addr == 0 {
+			return nil, malformed("can not set breakpoint at address 0x0")
+		}
+		return &AddrLocationSpec{uint64(addr)}, nil
+
+	default:
+		v := strings.SplitN(rest, ":", 2)
+
+		spec := &NormalLocationSpec{}
+
+		spec.Base = v[0]
+		spec.FuncBase = parseFuncLocationSpec(spec.Base)
+
+		if len(v) < 2 {
+			spec.LineOffset = -1
+			return spec, nil
+		}
+
+		rest = v[1]
+
+		var err error
+		spec.LineOffset, err = strconv.Atoi(rest)
+		if err != nil || spec.LineOffset < 0 {
+			return nil, malformed("line offset negative or not a number")
+		}
+
+		return spec, nil
+	}
+}
+
+func readRegex(in string) (rx string, rest string) {
+	out := make([]rune, 0, len(in))
+	escaped := false
+	for i, ch := range in {
+		if escaped {
+			if ch == '/' {
+				out = append(out, '/')
+			} else {
+				out = append(out, '\\')
+				out = append(out, ch)
+			}
+			escaped = false
+		} else {
+			switch ch {
+			case '\\':
+				escaped = true
+			case '/':
+				return string(out), in[i:]
+			default:
+				out = append(out, ch)
+			}
+		}
+	}
+
+	return string(out), ""
+}
+
+func parseFuncLocationSpec(in string) *FuncLocationSpec {
+	if strings.Index(in, "/") >= 0 {
+		return nil
+	}
+
+	v := strings.Split(in, ".")
+
+	var spec FuncLocationSpec
+
+	switch len(v) {
+	case 1:
+		spec.BaseName = v[0]
+
+	case 2:
+		spec.BaseName = v[1]
+		r := stripReceiverDecoration(v[0])
+		if r != v[0] {
+			spec.ReceiverName = r
+		} else {
+			spec.PackageOrReceiverName = r
+		}
+
+	case 3:
+		spec.BaseName = v[2]
+		spec.ReceiverName = stripReceiverDecoration(v[1])
+		spec.PackageName = stripReceiverDecoration(v[0])
+
+	default:
+		return nil
+	}
+
+	return &spec
+}
+
+func stripReceiverDecoration(in string) string {
+	if len(in) < 3 {
+		return in
+	}
+	if (in[0] != '(') || (in[1] != '*') || (in[len(in)-1] != ')') {
+		return in
+	}
+
+	return in[2 : len(in)-1]
+}
+
+func (spec *FuncLocationSpec) Match(sym *gosym.Sym) bool {
+	if spec.BaseName != sym.BaseName() {
+		return false
+	}
+
+	recv := stripReceiverDecoration(sym.ReceiverName())
+
+	if spec.ReceiverName != "" && spec.ReceiverName != recv {
+		return false
+	}
+
+	if spec.PackageName != "" && spec.PackageName != sym.PackageName() {
+		return false
+	}
+
+	if spec.PackageOrReceiverName != "" && spec.PackageOrReceiverName != sym.PackageName() && spec.PackageOrReceiverName != recv {
+		return false
+	}
+
+	return true
+}
+
+func (loc *RegexLocationSpec) Find(d *Debugger, locStr string) ([]api.Location, error) {
+	funcs := d.process.Funcs()
+	r := make([]api.Location, 0, len(funcs))
```

Comment:
Allocate this after we get `matches` back and use `len(matches)` instead of `len(funcs)`.